pkg/cloudprovider/providers/openstack/openstack_volumes.go

```diff
@@ -317,8 +317,20 @@ func (os *OpenStack) AttachDisk(instanceID, volumeID string) (string, error) {
 		if instanceID == volume.AttachedServerId {
 			glog.V(4).Infof("Disk %s is already attached to instance %s", volumeID, instanceID)
 			return volume.ID, nil
+		} else {
+			attachErr := fmt.Sprintf("disk %s is attached to a different instance (%s)", volumeID, volume.AttachedServerId)
+			glog.V(4).Infof(attachErr)
+
+			//attachedInstance, ierr := os.getInstanceByID(instanceID)
+
+			//nodeName := mapServerToNodeName(attachedInstance)
+
+			// should we use os.GetDevicePathBySerialId(volumeID) or volume.AttachedDevice see issue #33128
```

Comment:
I am now thinking could #50004 happen because of #33128? If devices are detached by devicepath, that might cause that problem that drives are not detached at all?

Correct way of doing danglingerror is maybe use volume.AttachDevice, but if that returns incorrect information we have have quite big problem. Instead if we insert "/dev/disk/by-id/", id to danglingerror, it might mean that it does not work at all?

Then second problem is that in openstack we do not have function like getinstancebyid. Where I can get nodename using instanceid without querying openstack api? That is of course the last option, but I would not like go there.

Reply:
@kiall because you fixed that devicepath thing, could you check comment above ^.

Reply:
>  Where I can get nodename using instanceid without querying openstack api? That is of course the last option, but I would not like go there.

Why would you not want to go there? It seems pretty trivial to get `nodeName` from instance-id using API. Something like:

```
server, err := servers.Get(i.compute, instanceID).Extract()
```

cc @NickrenREN

Reply:
Well, if there is option to use existing data or load data from api always. My choice is use existing data. Like in this case nodename is quite static info - we could cache it.

But if there isnt option to use existing data then i must use api.

Still question is about that devicepath things. From my perspective it might be that these cinder detach issues might be related to incorrect paths. Cinder in openstack says like /dev/vdb but then when you go to machine it is located in /dev/vdc. I have seen problems like this in our openstack. However, i think we should here make kube trust cinder devicepath. Then openstack guys should fix their side. What you think?

Reply:
I do not think detaching is related to device path problem because volume detach code doesn't appear to be using device path at all. 

If somehow detach is affected by incorrect device path, then it could be because umount is not working properly. I am yet to properly look into bugs linked in this PR to verify if that was the case.